sdk/lib/queue/queue.cc

```diff
@@ -0,0 +1,437 @@
+#include <cheri.hh>
+#include <errno.h>
+#include <locks.hh>
+#include <queue.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <timeout.h>
+#include <type_traits>
+
+using namespace CHERI;
+using cheriot::atomic;
+
+using Debug = ConditionalDebug<true, "Queue library">;
+
+#ifdef __cplusplus
+using MessageQueueCounter = atomic<uint32_t>;
+#else
+typedef uint32_t MessageQueueCounter;
+#endif
+
+#include <assembly-helpers.h>
+
+namespace
+{
+	/**
+	 * Helpers for the queue.  The queue uses two counters that wrap on double
```

Comment:
Might be worth making https://github.com/microsoft/cheriot-rtos/blob/main/sdk/include/ds/ring_buffer.h use that trick rather than the separate `bool empty` and then use that here?

Reply:
Please do.  I mostly wanted you to check that this makes sense.  I haven't found anywhere else that uses this trick from a quick skim of things explaining how to design a ring buffer, but I can't quite believe I invented it because it seems too obvious.  I generalised this from the power-of-two version.  15 years ago I worked through a proof of why that worked because I didn't initially believe it, and I think it is basically the same for this.  The nice thing about this is that it requires one bit of extra state (which is otherwise padding because we don't have variable-bit-width memory) and doesn't require any divide operations (they're 15 cycles on Ibex).  That seems too nice to be real, so I suspect I got some maths wrong.